### PR TITLE
New downsample method

### PR DESCRIPTION
**Pull request by chris-langfield**
The final stage of adding the new downsampling method.

This relates to issues

#140, #608, #566

The method uses the new `crop_pad_2d` method in the frequency domain and is tested by new downsampling tests that do not rely on hardcoded results

### CODE REVIEW

**Comment by chris-langfield** [comment]
Using `centered_fft2` fixes the issues from #608. The new downsampling method now passes the logical tests introduced in #567 

It is worth noting that `Volume.downsample()` is not currently replaced with this new method. 
It uses the related but not identical method `aspire.image.downsample()`. 

Although this method uses a similar strategy, I believe there is a conversation to be had about how to best go about vetting and possibly improving this 3D downsampling method.  (I made an issue: #617)

File: src/aspire/image/image.py [hunk 1]
```diff
@@ -209,30 +208,16 @@ def downsample(self, ds_res):
...
-
-        for s in range(im_ds.shape[0]):
-            interpolator = RegularGridInterpolator(
-                (x, y), im[s], bounds_error=False, fill_value=0
-            )
-            im_ds[s] = interpolator(np.dstack([grid_ds["y"], grid_ds["x"]]))
-
-        return Image(im_ds)
+        # compute FT with centered 0-frequency
+        fx = np.array([fft.centered_fft2(self.data[i]) for i in range(self.n_images)])
```
Comment (j-c-c) [inline src/aspire/image/image.py]: Should be able to just do `fx = fft.centered_fft2(self.data)`. I don't think the loop is necessary. Same below for `ifft`.

File: src/aspire/image/image.py [hunk 2]
```diff
@@ -209,30 +208,16 @@ def downsample(self, ds_res):
...
+        # crop 2D Fourier transform for each image
+        crop_fx = np.array([crop_pad_2d(fx[i], ds_res) for i in range(self.n_images)])
+        # take back to real space, discard complex part, and scale
+        out = np.real(
+            np.array([fft.centered_ifft2(crop_fx[i]) for i in range(self.n_images)])
+        ) * (ds_res**2 / self.res**2)
```
Comment (j-c-c) [inline src/aspire/image/image.py]: Maybe just `out = np.real(fft.centered_ifft2(crop_fx)) * (ds_res**2 / self.res**2)`?

**Review by j-c-c — CHANGES_REQUESTED** [review]
Looks good! Couple small things.

**Review by j-c-c — CHANGES_REQUESTED** [review]
Looks good! Couple small things.

File: src/aspire/image/image.py (hunk 2 above)
Comment (chris-langfield) [reply src/aspire/image/image.py]: I prefer the current set up (the `return Image(out)` echoes the pattern used in other parts of the code)

Comment (j-c-c) [reply src/aspire/image/image.py]: This was just in reference to the unnecessary list comprehension. Agree with matching other parts of the code for returning an `Image`.

**Review by j-c-c — APPROVED** [review]
Looks good!

File: src/aspire/image/image.py (hunk 2 above)
Comment (chris-langfield) [reply src/aspire/image/image.py]: Oh totally misunderstood. Your comment is absolutely right and I made the change thanks!

**Review by janden — APPROVED** [review]
Looks good.

**Pull request merged**